### ISSUE TITLE
Race condition when sending lots of messages

### ISSUE THREAD

**Issue by yschroeder** [eclipse/paho.mqtt.python#376]
I try to send a lot of MQTT messages as fast as possible (I know MQTT might not be the correct protocol for this...). I am using the following code snippet:

```python
from paho.mqtt.client import Client as MQTTClient
from paho.mqtt.client import error_string

mqttc = MQTTClient()
mqttc.connect_async('localhost', 1883, 10)
mqttc.loop_start()

# adding these two lines, it always works
#info = mqttc.publish("test-topic", payload="online", qos=2, retain=False)
#info.wait_for_publish()

for i in range(300):
    message = "{}".format(i)
    info = mqttc.publish("test-topic", payload=message, qos=2, retain=False)
    print(message, error_string(info.rc))

# adding this line also makes it work, but data is being send "chunky"
# it sends 120 messages at a time and then waits quite some time
# until the next chunk is send
#info.wait_for_publish()
mqttc.loop_stop()
mqttc.disconnect()
```

What I experience is, that when running this script, occasionally it does not send messages at all (most of the time it works perfectly). I believe in the error case, the client never manages to connect in time and the loop is stopped and MQTT disconnected before any message was being able to leave the client.

I want all of the messages being send before the connection is closed. Do I have to track this myself by saving all the `MQTTMessageInfo` objects and check if all of them have been sent?

Waiting for the last message to be sent helps (see the example). However, the messages are being send in chunks of 120 messages with 20 seconds delay in between. I suspect this to be due to message queue limits in Mosquitto (20 inflight messages with QoS > 0 and 100 queued messages maximum). However, why does the client wait 20 seconds before trying to send the next message?

Tl;DR: I was expecting that the loop only stops when all QoS messages were sent correctly. What do I need to configure to get rid of the 20 second delay once the queue at the broker is full and make it retry earlier?

**Comment by yschroeder** [comment]
I just figured out that the message retry time is set to 20 seconds by default. On PyPI the documentation is wrong: https://pypi.org/project/paho-mqtt/#message-retry-set

It says 5 seconds there. Setting it to a lower values with my problem.

**Comment by yschroeder** [comment]
Part of the problem seems to be related to #196. I think the library should be able to tell me if all QoS > 0 messages have been sent correctly.

How about adding a function e.g. `message_queue_empty()` that returns `True` if all QoS > 0 messages have been sent? When calling `loop_stop(force=False)` it could wait for this to be `True` and only join the thread once the messages are sent. With `loop_stop(force=True)` it would not wait for this condition, but exit immediately.

**Comment by vrst37** [comment]
Hi, does [this](https://github.com/eclipse/paho.mqtt.python/blob/a8b2b4d039571e9bd873dd550a1dddabc6d3cb61/src/paho/mqtt/client.py#L1449) fulfill your requirement?

**Comment by Mithrandir2k18** [comment]
I also encountered a race condition. My scenario was a sensor that is sending data at 1Hz(calls the publish function from an asyncio-thread), is syncronized via Ping/Pong/Offset Messages(the ping is replied to in the on_message callback) at 3Hz and sends a Heartbeat at 0.25Hz from another thread.

The race condition would sometimes occur after 10 seconds, sometimes after 20 minutes at a steady traffic.

**Comment by MattBrittan** [comment]
I believe that @vst37 has provided some options and there has been no response so I'm going to close this issue (as its very old and the library has changed in the interim). If anyone is still having issues please feel free to reopen.

@Mithrandir2k18 - It's not really clear if your issue is the same as OP's; if you are still seeing this please rais a new issue (but please provide a minimal reproducible example if at all possible).

Ref the original issue; I think it's important to note that `connect_async` returns immediately (the connection process happens in the network loop). As such it's possible that you will get to the `loop_stop` before the connection is actually up (at QOS1+ messages are queued if the connection is down) and a bit more work is needed to ensure messages actually go out.

**Comment by Mithrandir2k18** [comment]
@MattBrittan This issue is rather old and I don't want to be rude, but I don't really care anymore. We switched to beerfactory/hbmqtt and that didn't cause any race conditions in our usecase.

What I described is a setup with 4.25 messages per second coming from 3 different threads, setting this up is trivial. We observed that sending a message from a callback thread (subscriber of a ping topic publishes a pong on a pong topic) *could* trigger the race-condition sometimes.

Sorry but that's all the info I can give at this moment - and I don't know if it's still relevant anyway as the project in question was completed years ago.
